rholang/src/test/scala/coop/rchain/rholang/ProcGen.scala

```diff
@@ -0,0 +1,247 @@
+          go(localNames ++ newNames, n.proc_)
+
+      }
+    go(Set.empty[String], p)
+  }
+}
+
+object ProcGen {
+  import tools._
+
+  case class State(height: Int, usedNames: Set[String]) {
+    def decrementHeight: State                   = this.copy(height = height - 1)
+    def addNames(names: Iterable[String]): State = this.copy(usedNames = usedNames ++ names)
+  }
+
+  private def pparGen(state: State): Gen[PPar] =
+    for {
+      p1 <- procGen(processContextProcs, state.decrementHeight)
+      p2 <- procGen(processContextProcs, state.decrementHeight)
+    } yield new PPar(p1, p2)
+
+  private def pvarGen(state: State): Gen[Proc] =
+    if (state.usedNames.isEmpty)
+      pnilGen
+    else
+      procVarVarGen(Gen.oneOf(state.usedNames.toSeq)).map(new PVar(_))
+
+  private def procVarVarGen(identifierGen: Gen[String]): Gen[ProcVarVar] =
+    identifierGen.map(new ProcVarVar(_))
+
+  private def pgroundGen(state: State): Gen[Proc] = {
+    lazy val groundIntGen = arbitrary[Long]
+      .map((n: Long) => {
+        if (n > 0)
+          new PGround(new GroundInt(n.toString))
+        else new PNeg(new PGround(new GroundInt((-n).toString)))
+      })
+    lazy val groundBoolGen =
+      Gen.oneOf(new BoolFalse(), new BoolTrue()).map(b => new PGround(new GroundBool(b)))
+    lazy val groundStringGen =
+      Arbitrary.arbString.arbitrary.map(s => new PGround(new GroundString(withQuotes('"')(s))))
+    lazy val groundUriGen =
+      uriGen.map(s => new PGround(new GroundUri(s)))
+
+    Gen.oneOf(
+      groundIntGen,
+      groundBoolGen,
+      groundStringGen,
+      groundUriGen
+    )
+  }
+
+  private def nameQuoteGen(state: State): Gen[NameQuote] =
+    procGen(allProcs, state.decrementHeight).map(new NameQuote(_))
+  private def nameGen(state: State): Gen[Name] =
+    nameQuoteGen(state)
+
+  lazy val sendSingleGen: Gen[SendSingle]     = Gen.const(new SendSingle())
+  lazy val sendMultipleGen: Gen[SendMultiple] = Gen.const(new SendMultiple())
+  lazy val sendGen: Gen[Send] = Gen.oneOf(
+    sendSingleGen,
+    sendMultipleGen
+  )
+
+  private def psendGen(state: State): Gen[PSend] =
+    for {
+      name     <- nameGen(state.decrementHeight)
+      send     <- sendGen
+      listProc <- Gen.listOf(procGen(allProcs, state.decrementHeight))
+    } yield new PSend(name, send, seqToJavaCollection[ListProc, Proc](listProc))
+
+  lazy val pnilGen: Gen[PNil] = Gen.const(new PNil())
+
+  private def nameDeclSimplGen: Gen[NameDeclSimpl] = identifierGen.map(new NameDeclSimpl(_))
+
+  lazy val nameDeclUrnGen: Gen[NameDeclUrn] =
+    for {
+      name <- identifierGen
+      uri  <- uriGen
+    } yield new NameDeclUrn(name, uri)
+
+  lazy val nameDeclGen: Gen[NameDecl] = Gen.oneOf(nameDeclSimplGen, nameDeclUrnGen)
+
+  private def pnewGen(state: State): Gen[PNew] = {
+    val newState = state.decrementHeight
+
+    for {
+      seqNameDecl <- Gen.nonEmptyListOf(nameDeclSimplGen)
+      names       = extractNames(seqNameDecl)
```

Comment:
Why not just `names <- Gen.nonEmptyListOf(identifierGen)` ?

Reply:
If I do it your way then I will have to create a generator of `NameDecl` which is allowed to only pick one of a set of pregenerated names which is more difficult because this generator will have to deal with a global state i.e. should know about its previous runs in the given context.

The way I implemented it, I'm generating random `NameDecl` names very simple with a small risk of collisions and then extract the names. Also note that this function is required anyway in the shrinker too.

Reply:
I don't see how this is the case given

```scala
private def nameDeclSimplGen: Gen[NameDeclSimpl] = identifierGen.map(new NameDeclSimpl(_))
```

So basically lines 166-167 generate a `nonEmptyListOf(identifierGen)`, wrap them in `NameDeclSimpl(_)` and unwrap them using `extractNames`. Unless I'm missing something, it's an unneeded inefficiency.

Reply:
I'm not neglecting the purpose of `extractNames`, just the validity of its usage here :)

Reply:
this was clarified over a zoom discussion